x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java

```diff
@@ -72,26 +79,49 @@ public TransportLoadTrainedModelPackage(
         this.client = new OriginSettingClient(client, ML_ORIGIN);
     }
 
+    @Override
+    protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+        return null;
+    }
+
     @Override
     protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener)
         throws Exception {
-        String modelId = request.getModelId();
-        ModelPackageConfig packageConfig = request.getModelPackageConfig();
-        ModelImporter modelImporter = new ModelImporter(client, modelId, packageConfig);
+        CancellableTask downloadTask = createDownloadTask(request);
+        ParentTaskAssigningClient parentTaskAssigningClient = getParentTaskAssigningClient(downloadTask);
+
+        ModelImporter modelImporter = new ModelImporter(
+            parentTaskAssigningClient,
+            request.getModelId(),
+            request.getModelPackageConfig(),
+            downloadTask
+        );
 
         threadPool.executor(MachineLearningPackageLoader.UTILITY_THREAD_POOL_NAME)
-            .execute(() -> importModel(client, request, modelImporter, listener));
+            .execute(() -> importModel(client, taskManager, request, modelImporter, listener, downloadTask));
 
         if (request.isWaitForCompletion() == false) {
             listener.onResponse(AcknowledgedResponse.TRUE);
         }
     }
 
+    private ParentTaskAssigningClient getParentTaskAssigningClient(Task originTask) {
+        var parentTaskId = new TaskId(clusterService.localNode().getId(), originTask.getId());
+        return new ParentTaskAssigningClient(client, parentTaskId);
+    }
+
     /**
      * This is package scope so that we can test the logic directly.
      * This should only be called from the masterOperation method and the tests
      */
-    static void importModel(Client client, Request request, ModelImporter modelImporter, ActionListener<AcknowledgedResponse> listener) {
+    static void importModel(
+        Client client,
+        TaskManager taskManager,
+        Request request,
+        ModelImporter modelImporter,
+        ActionListener<AcknowledgedResponse> listener,
+        Task task
+    ) {
         String modelId = request.getModelId();
         final AtomicReference<Exception> exceptionRef = new AtomicReference<>();
```

Comment:
as `importModel` takes over ownership of the task these 2 lines should be part of `try {`